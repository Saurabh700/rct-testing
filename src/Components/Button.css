.button {
  color: white;
  cursor: pointer;
  padding: 10px;
  margin-top: 1rem;
  border-radius: 5px;
}

.blue {
  background-color: blue;
}

.green {
  background-color: greenyellow;
}

.ghost {
  border: none;
}

.outline {
  background-color: white;
  border: 1px solid green;
  color: black;
}

.crimson {
  background-color: crimson;
}
