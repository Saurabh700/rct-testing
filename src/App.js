import { useState } from "react";
import "./App.css";
import Button from "./Components/Button";
import Counter from "./Components/Counter";

function App() {
  const [theme, setTheme] = useState("light");
  return (
    <div className="App">
      <h1>Theme is {theme}</h1>
      <Button
        colorScheme={"crimson"}
        variant={"outline"}
        // onClick={() => setTheme(theme === "light" ? "dark" : "light")}
        onClick={() =>
          theme === "light" ? setTheme("dark") : setTheme("light")
        }
      >
        Click Me
      </Button>

      <hr />

      <Counter />
    </div>
  );
}

export default App;
