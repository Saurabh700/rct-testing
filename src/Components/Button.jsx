import React from "react";
import "./Button.css";

const Button = ({ children, colorScheme, variant, onClick }) => {
  return (
    // we have replicated button component of chakra ui
    <button
      data-testid="customButton"
      className={`button ${colorScheme || "blue"} ${variant || "ghost"}`}
      onClick={onClick}
    >
      {children}
    </button>
  );
};

export default Button;
