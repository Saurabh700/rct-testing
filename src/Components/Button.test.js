import Button from "./Button";
import { render, screen } from "@testing-library/react";
import App from "../App";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

// instead of userEvent we can also use fireEvent

// to use snapshot we need to install -> npm install react-test-renderer

// in place of test we can also use "it" both works the same

// here i am testing that the custom Button component that i have created is actually getting rendered on the screen or not

describe("testing the custom button", () => {
  // not preferred because there can be more elements with text "Click Me"
  test("should be present in the dom", () => {
    render(<Button>Click Me</Button>);
    // screen.debug();
    let button = screen.getByText("Click Me");

    expect(button).toBeInTheDocument();
  });

  //   Preferred because there can be only one elem with a id
  it("should have a button with text click me", () => {
    render(<App></App>); //virtually
    // screen.debug();
    let button = screen.getByTestId("customButton");
    expect(button).toBeInTheDocument();
  });

  //   Empty dom element =>  <div></div>

  it("should render without any children", () => {
    render(<Button></Button>);
    // render(<Button>Click</Button>); this will throw an error
    screen.debug();
    let button = screen.getByTestId("customButton");
    expect(button).toBeEmptyDOMElement();
  });

  it("should toggle the theme", () => {
    render(<App></App>);
    // this is the flow of events in testing
    let element = screen.getByText("Theme is light");
    expect(element).toHaveTextContent("light");
    // pehle esa tha

    let button = screen.getByTestId("customButton");
    userEvent.click(button);
    // jab button par click karengey

    expect(element).toHaveTextContent("dark");
    userEvent.click(button);
    expect(element).toHaveTextContent("light");
    // toh esa ho jayega
  });

  //   to check ki click karne par function execute ho rha hai ki nahi
  it("should call the given function", () => {
    const mockFn = jest.fn();
    render(<Button onClick={mockFn}>Click Me</Button>);
    let button = screen.getByTestId("customButton");
    userEvent.click(button);
    userEvent.click(button);
    // expect(mockFn).toBeCalled();
    expect(mockFn).toBeCalledTimes(2);
  });

  //   snapshot
  it("should match the snapshot", () => {
    const tree = renderer
      .create(
        <Button colorScheme={"green"} variant={"outline"}>
          Increment Counter
        </Button>
      )
      .toJSON(); //instead of .toJSON() we can also use .toTree()
    expect(tree).toMatchSnapshot();
  });

  //   if anything changes then snapshot fails -->but we want tokeep that change then in terminal press w and u --> we will get this option only when if the testcase fails

  //   this will create a snapshot in __snapshot__ folder
});
