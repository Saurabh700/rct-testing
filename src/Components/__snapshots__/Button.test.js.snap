// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the custom button should match the snapshot 1`] = `
<button
  className="button green outline"
  data-testid="customButton"
>
  Increment Counter
</button>
`;
